### DISCUSSION TITLE
Proof Wallet, a minimalist multisig cosigner wallet

### DISCUSSION THREAD

**Discussion by hodlwave** [BlockchainCommons/Gordian-Developer-Community#23] [Show and tell]
Hi airgapped wallet enthusiasts,

I've been working on [Proof Wallet](https://github.com/hodlwave/proof-wallet/), a multisig cold storage solution inspired by the Glacier Protocol, which aims to improve on some of its limitations without compromising on security.

Its features include...
• Interoperable with other offline signers (hardware / software wallets)
• Minimal wallet (~1200 lines) that relies on Bitcoin Core to do heavy lifting
• Uses established standards: BIP39, BIP32, BIP174, output descriptors
• Few external dependencies (qrencode & zbarcam for shuttling data via QR codes & python-mnemonic for BIP39)
• avoids hardware wallet supply chain risk
• Users can sign transactions sequentially so private keys can always be physically & electronically separated
• Full test coverage

I was motivated to make Proof Wallet for a few reasons
• Glacier Protocol has many glaring deficiencies, which have likely contributed to the project's total abandonment
• Lack of hardware wallet multisig support, and other security hardware wallet issues such as supply chain risk, lack of proper airgaps, and poor or nonexistent code review

I was hoping to get some code review for Proof Wallet as well as suggestions/advice on how to make it a legitimate option Bitcoiners can utilize to custody serious funds. For example, one glaring issue is the lack of documentation (at least compared to Glacier Protocol's exhaustive instructions) although I did add a quickstart to the repo recently to help folks, who'd like to try it out on testnet.

**Comment by Fonta1n3** [comment]
I had tested awhile ago and really was pleased with it. Will definitely be testing your latest release along with the descriptor compatibility very soon and give feedback here. In my opinion this is an excellent tool. 

I think adding support for `crypto-psbt` UR would be really nice for transmitting psbts via QR code. There is a python library for UR iirc.

**Comment by hodlwave** [comment]
Hi @Fonta1n3,

I think support for UR could be a good idea; I'm going to research it further and probably create a branch that uses it. Currently, Proof Wallet encodes its QR codes in plaintext (the raw output descriptor string and base64 encoded psbts). I'm rather arbitrarily limiting each signed psbt QR code to 1000 bytes so for larger psbts users have to do more awkward ad hoc manual work assembling strings together.

Thanks for the initial feedback, and I look forward to hearing more when you try out the new descriptor-based functionality :)

**Reply by ChristopherA** [reply]
UR supports multiple QRs & animated QRs for PSBTs that are large.
